pkg/webhook/certs.go

```diff
@@ -0,0 +1,369 @@
+package webhook
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"math/big"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"time"
+)
+
+const (
+	caName     = "gatekeeper-ca"
+	namespace  = "gatekeeper-system"
+	service    = "gatekeeper-webhook-service"
+	certName   = "tls.crt"
+	keyName    = "tls.key"
+	caCertName = "ca.crt"
+	caKeyName  = "ca.key"
+)
+
+var crLog = logf.Log.WithName("cert-rotation")
+
+var (
+	secretKey = types.NamespacedName{
+		Namespace: namespace,
+		Name:      "gatekeeper-webhook-server-cert",
+	}
+	// DNS name is <service name>.<namespace>.svc
+	DNSName = fmt.Sprintf("%s.%s.svc", service, namespace)
+)
+
+var _ manager.Runnable = &certRotator{}
+
+func NewRotator(mgr manager.Manager) (*certRotator, error) {
+	// Use a new client so we are unaffected by the cache sync kill signal
+	cli, err := client.New(mgr.GetConfig(), client.Options{Scheme: mgr.GetScheme(), Mapper: mgr.GetRESTMapper()})
+	if err != nil {
+		return nil, err
+	}
+	return &certRotator{client: cli}, nil
+}
+
+type certRotator struct {
+	client client.Client
+}
+
+func (cr *certRotator) Start(stop <-chan (struct{})) error {
+	// explicitly rotate on the first round so that the certificate
+	// can be bootstrapped, otherwise manager exits before a cert can be written
+	crLog.Info("starting cert rotator controller")
+	defer crLog.Info("stopping cert rotator controller")
+	if restart, err := cr.refreshCertIfNeeded(); err != nil {
+		crLog.Error(err, "could not refresh cert on startup")
+		return errors.Wrap(err, "could not refresh cert on startup")
+	} else if restart {
+		crLog.Info("certs refreshed, restarting server")
+		return nil
+	}
+	ticker := time.NewTicker(12 * time.Hour)
```

Comment:
Can we make this duration a const or configurable?

Reply:
Done, changed to constant.